File: ui/src-tauri/src/commands/secrets.rs
use super::{bridge_call, CommandResponse};
use crate::bridge::BridgeManager;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::sync::Arc;
use tauri::State;

#[derive(Debug, Serialize, Deserialize)]
pub struct SecretInfo {
    pub name: String,
    pub source: String, // "1password" | "env" | "docker"
    pub mapped_to: Vec<String>,
    pub last_synced: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SecretsList {
    pub source: String,
    pub environment: String,
    pub secrets: Vec<SecretInfo>,
    pub mapped_count: u32,
    pub total_count: u32,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SyncResult {
    pub success: bool,
    pub synced: u32,
    pub failed: u32,
    pub dry_run: bool,
    pub results: Vec<SyncStepResult>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SyncStepResult {
    pub secret: String,
    pub from_source: String,
    pub to_target: String,
    pub status: String, // "synced" | "failed" | "would_sync" | "skipped"
    pub error: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct VerifyResult {
    pub success: bool,
    pub in_sync: u32,
    pub out_of_sync: u32,
    pub results: Vec<VerifyStepResult>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct VerifyStepResult {
    pub secret: String,
    pub status: String, // "in_sync" | "out_of_sync" | "missing" | "error"
    pub details: Option<String>,
}

/// List secrets
#[tauri::command]
pub fn list_secrets(
    bridge: State<Arc<BridgeManager>>,
    project_path: String,
    environment: Option<String>,
    source: Option<String>,
) -> CommandResponse<Value> {
    match bridge_call(
        &bridge,
        "secrets.list",
        Some(json!({
            "path": project_path,
            "environment": environment,
            "source": source
        })),
    ) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

/// Sync secrets between sources
#[tauri::command]
pub fn sync_secrets(
    bridge: State<Arc<BridgeManager>>,
    project_path: String,
    from_source: String,
    to_target: String,
    environment: Option<String>,
    dry_run: bool,
) -> CommandResponse<Value> {
    match bridge_call(
        &bridge,
        "secrets.sync",
        Some(json!({
            "path": project_path,
            "from": from_source,
            "to": to_target,
            "environment": environment,
            "dry_run": dry_run
        })),
    ) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

/// Verify secrets are in sync
#[tauri::command]
pub fn verify_secrets(
    bridge: State<Arc<BridgeManager>>,
    project_path: String,
    environment: Option<String>,
) -> CommandResponse<Value> {
    match bridge_call(
        &bridge,
        "secrets.verify",
        Some(json!({
            "path": project_path,
            "environment": environment
        })),
    ) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

/// Export secrets to file or env format
#[tauri::command]
pub fn export_secrets(
    bridge: State<Arc<BridgeManager>>,
    project_path: String,
    environment: String,
    format: String, // "env" | "json" | "yaml"
) -> CommandResponse<Value> {
    match bridge_call(
        &bridge,
        "secrets.export",
        Some(json!({
            "path": project_path,
            "environment": environment,
            "format": format
        })),
    ) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

/// Get available secret providers
#[tauri::command]
pub fn get_secret_providers(bridge: State<Arc<BridgeManager>>) -> CommandResponse<Value> {
    match bridge_call(&bridge, "secrets.providers", None) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

File: ui/src/api/secrets.ts
/**
 * Secrets API
 */

import { invoke } from '@tauri-apps/api/core';
import type { CommandResponse } from '../types';

export interface SecretInfo {
  name: string;
  source: string; // "1password" | "env" | "docker"
  mapped_to: string[];
  last_synced: string | null;
}

export interface SecretsList {
  source: string;
  environment: string;
  secrets: SecretInfo[];
  mapped_count: number;
  total_count: number;
  error?: string;
}

export interface SyncResult {
  success: boolean;
  synced: number;
  failed: number;
  dry_run: boolean;
  results: Array<{
    secret: string;
    from_source: string;
    to_target: string;
    status: string;
    error: string | null;
  }>;
}

export interface VerifyResult {
  success: boolean;
  in_sync: number;
  out_of_sync: number;
  results: Array<{
    secret: string;
    status: string;
    details: string | null;
  }>;
}

export interface SecretProvider {
  name: string;
  type: string;
  available: boolean;
  authenticated: boolean;
}

/**
 * List secrets
 */
export async function listSecrets(
  projectPath: string,
  environment?: string,
  source?: string
): Promise<SecretsList> {
  const response = await invoke<CommandResponse<SecretsList>>('list_secrets', {
    projectPath,
    environment,
    source,
  });
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to list secrets');
  }
  return response.data;
}

/**
 * Sync secrets between sources
 */
export async function syncSecrets(
  projectPath: string,
  fromSource: string,
  toTarget: string,
  environment?: string,
  dryRun: boolean = false
): Promise<SyncResult> {
  const response = await invoke<CommandResponse<SyncResult>>('sync_secrets', {
    projectPath,
    fromSource,
    toTarget,
    environment,
    dryRun,
  });
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to sync secrets');
  }
  return response.data;
}

/**
 * Verify secrets are in sync
 */
export async function verifySecrets(
  projectPath: string,
  environment?: string
): Promise<VerifyResult> {
  const response = await invoke<CommandResponse<VerifyResult>>('verify_secrets', {
    projectPath,
    environment,
  });
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to verify secrets');
  }
  return response.data;
}

/**
 * Export secrets to file or env format
 */
export async function exportSecrets(
  projectPath: string,
  environment: string,
  format: 'env' | 'json' | 'yaml' = 'env'
): Promise<{ content: string; format: string }> {
  const response = await invoke<CommandResponse<{ content: string; format: string }>>(
    'export_secrets',
    {
      projectPath,
      environment,
      format,
    }
  );
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to export secrets');
  }
  return response.data;
}

/**
 * Get available secret providers
 */
export async function getSecretProviders(): Promise<{ providers: SecretProvider[] }> {
  const response = await invoke<CommandResponse<{ providers: SecretProvider[] }>>(
    'get_secret_providers'
  );
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to get secret providers');
  }
  return response.data;
}
